File: CodingSample/DotNet/Enterprise/AspNetCoreMvc/ControllerTest/DemoApp2/Tourism/Models/SiteModel.cs
using Tourism.Data;

namespace Tourism.Models;

public class SiteModel
{
    public IEnumerable<Visitor> GetVisitors()
    {
        using var site = new SiteDbContext();
        var selection = from t in site.Travelers
                        where t.Id.Length > 3
                        select new Visitor 
                        {
                            Name = t.Id,
                            Visits = t.Tours.Count,
                            Recent = t.Tours.Max(e => e.Checkin),
                            Stars = new string('*', t.Rating)
                        };
        return selection.ToList();
    }

    public void AcceptVisit(string visitorId, int visitorRating)
    {
        using var site = new SiteDbContext();
        var traveler = site.Travelers.Find(visitorId);
        if(traveler is null)
        {
            traveler = new Traveler { Id = visitorId };
            site.Travelers.Add(traveler);
        }
        traveler.Tours.Add(new Trip());
        traveler.Rating = visitorRating;
        site.SaveChanges();
    }
}

File: CodingSample/Java/Advanced/JPA/EntityTest/demo/src/main/java/app/tourism/models/SiteModel.java
package app.tourism.models;

import app.tourism.data.TravelerEntity;
import app.tourism.data.TripEntity;
import jakarta.persistence.EntityManagerFactory;
import jakarta.persistence.Persistence;

public class SiteModel {
    
    private static EntityManagerFactory emf = Persistence.createEntityManagerFactory("app.data");

    public Iterable<Visitor> getVisitors() {
        try(var em = emf.createEntityManager()){
            var query = em.createQuery("SELECT t FROM TravelerEntity t WHERE LENGTH(t.id) > 3", TravelerEntity.class);
            return query.getResultStream()
                .map(Visitor::fromTraveler)
                .toList();
        }
    }

    public boolean acceptVisit(String visitorId, int visitorRating) {
        if(visitorId == null || visitorRating < 1 || visitorRating > 5)
            return false;
        try(var em = emf.createEntityManager()){
            var traveler = em.find(TravelerEntity.class, visitorId);
            if(traveler == null){
                traveler = new TravelerEntity();
                traveler.setId(visitorId);
            }
            traveler.setRating(visitorRating);
            var trip = new TripEntity();
            trip.setGuest(traveler);
            traveler.getTours().add(trip);
            var tx = em.getTransaction();
            tx.begin();
            em.persist(traveler);
            tx.commit();
            return true;
        }
    }

}
